google/cloud/spanner/value.cc

```diff
@@ -0,0 +1,156 @@
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+Value::Value(bool v) {
+  type_.set_code(google::spanner::v1::TypeCode::BOOL);
+  value_.set_bool_value(v);
+}
+
+Value::Value(std::int64_t v) {
+  type_.set_code(google::spanner::v1::TypeCode::INT64);
+  std::stringstream ss;
+  ss << std::dec << v;
+  value_.set_string_value(ss.str());
+}
+
+Value::Value(double v) {
+  type_.set_code(google::spanner::v1::TypeCode::FLOAT64);
+  if (std::isnan(v)) {
+    value_.set_string_value("NaN");
+  } else if (std::isinf(v)) {
+    value_.set_string_value(v < 0 ? "-Infinity" : "Infinity");
+  } else {
+    value_.set_number_value(v);
+  }
+}
+
+Value::Value(std::string v) {
+  type_.set_code(google::spanner::v1::TypeCode::STRING);
+  value_.set_string_value(std::move(v));
+}
+
+bool operator==(Value a, Value b) {
+  if (a.type_.code() != b.type_.code()) return false;
+  switch (a.type_.code()) {
+    case google::spanner::v1::TypeCode::BOOL:
+      return a.get<bool>() == b.get<bool>();
+    case google::spanner::v1::TypeCode::INT64:
+      return a.get<std::int64_t>() == b.get<std::int64_t>();
+    case google::spanner::v1::TypeCode::FLOAT64:
+      return a.get<double>() == b.get<double>();
+    case google::spanner::v1::TypeCode::STRING:
+      return a.get<std::string>() == b.get<std::string>();
+    default:
+      return false;
+  }
+}
+
+std::ostream& operator<<(std::ostream& os, Value v) {
+  auto const& code = v.type_.code();
+  auto const* const descriptor = google::spanner::v1::TypeCode_descriptor();
+  os << "Value{" << descriptor->FindValueByNumber(code)->name() << ":";
+  if (v.is_null()) return os << "(null)}";
+  switch (code) {
+    case google::spanner::v1::TypeCode::BOOL:
+      os << v.get<bool>();
+      break;
+    case google::spanner::v1::TypeCode::INT64:
+      os << v.get<std::int64_t>();
+      break;
+    case google::spanner::v1::TypeCode::FLOAT64:
+      os << v.get<double>();
+      break;
+    case google::spanner::v1::TypeCode::STRING:
+      os << v.get<std::string>();
+      break;
+  }
+  return os << "}";
+}
+
+bool Value::is_null() const {
+  return value_.kind_case() == google::protobuf::Value::kNullValue;
+}
+
+//
+// Value::IsType
+//
+
+bool Value::IsType(bool) const {
+  return type_.code() == google::spanner::v1::TypeCode::BOOL;
+}
+
+bool Value::IsType(std::int64_t) const {
+  return type_.code() == google::spanner::v1::TypeCode::INT64;
+}
+
+bool Value::IsType(double) const {
+  return type_.code() == google::spanner::v1::TypeCode::FLOAT64;
+}
+
+bool Value::IsType(std::string) const {
+  return type_.code() == google::spanner::v1::TypeCode::STRING;
+}
+
+//
+// Value::GetValue
+//
+
+bool Value::GetValue(bool) const {
+  if (is_null() || !is<bool>()) return {};  // XXX: Potentially UB or crash?
+  return value_.bool_value();
+}
+
+std::int64_t Value::GetValue(std::int64_t) const {
+  if (is_null() || !is<std::int64_t>())
+    return {};  // XXX: Potentially UB or crash?
```

Comment:
No change yet, mainly because I still have a question about whether or not this is actually UB. In the header docs for `get<T>()` I say

```cc
   * Returns the contained value if the specified type `T` matches the Value's
   * type and if the contained value is not null. Otherwise, a default
   * constructed `T` is returned.
```

So as of right now, it's explicitly not UB. But of course, I contradict myself (a little) by still including this comment in the implementation about whether this decision is correct and whether perhaps it *should* be UB.

FWIW, the current guidance about returning a default constructed `T` comes from following the behavior of protobuf's and their `oneof` behavior. There is no UB or crashes. Callers can test for null and check the type (with `is<T>()`), but if they then try to access the value the assumption is that they want the behavior of "return a default constructed T if it doesn't have a T". If they don't want that behavior, they can avoid it by checking for null and the type explicitly. 

Additionally, callers could call `v.get<optional<T>>()`, in which case it'll return a disengaged optional if the value was null.

.... sooooooo, do you have thoughts on this point?

---

A 2-step alternative design:

1. Change get to look like:

```cc
template <typename T>
StatusOr<T> get() const;
```
This gives us a non-lethal way to signal the error to the caller... but it's a little more cumbersom to use, because StatusOr.

2. Add an explicit conversion operator, that does NOT return StatusOr<>, and declare that using this interface is UB to cast to the wrong type or if the Value is null. E.g.,

```cc
  template <typename T>
  explicit operator T() const;
...
  Value v = ...;
  std::int64_t i = static_cast<std::int64_t>(v);
```
This lets callers avoid the StatusOr if they want to, but then they have to code around UB.

... I kinda like this alternative design.

Reply:
I like option (2) also.

Reply:
I'm surprised, but this static_cast<> idea will be a little harder than I thought, because of an oddity of optional (including std::optional). See https://godbolt.org/z/Zw2pWQ for an example.

Basically, `static_cast<std::optional<int>>(foo)` will end up calling foo's `operator int()`, **not** its `operator std::optional<int>`. The reason is because of optional's conversion operator precedence and the definition of how static_assert has to behave. But this was a surprise to me.

Ultimately, this means that from within our templated `explicit operator T()` we'll never see `T` as an optional, only the optional's contained value. And that would mean that we could not correctly convert to an `optional<T>` that is disengaged for "null" Value objects.

That *might* be the nail in the coffin for option (2). I'll look at again tomorrow when I'm rested, but I wanted to record these findings somewhere in the meantime.

Reply:
OK, I just published a new version of this CL. Call it option 2a :-) Here are the major differences.

1. There are no optionals in the API at all. The only way to construct a "null" Value is via a new factory function `Value::MakeNull<T>()`

2. `Value::get<T>` now returns a `StatusOr<T>`. This object will only have a non-OK status if the programmer passed the wrong T or if the Value was null. There is no potential UB when calling `Value::get`.

3. There's an `explicit operator T() const` operator, which callers can use to `static_cast` Value objects to the desired type. This operator requires that the caller verify on their own that the specified `T` is correct and that the Value is not null. Otherwise, the result is undefined. This is an alternative to calling `Value::get`. It might be convenient for callers writing generic code where using a cast is desirable.

PTAL and re-verify approval.